nautobot_golden_config/utilities/helper.py

```diff
@@ -139,3 +150,103 @@ def get_device_to_settings_map(queryset):
         if dynamic_group.exists():
             device_to_settings_map[device.id] = dynamic_group.first().golden_config_setting
     return device_to_settings_map
+
+
+def get_secret(
+    user: User,
+    obj_id: str,
+    obj_type: str,
+    secret_type: str,
+    secret_access_type: Optional[str] = SecretsGroupAccessTypeChoices.TYPE_GENERIC,
+) -> Optional[str]:
+    """Gets the secrets attached to an object based on an ORM relationship. To be used as a Jinja filter.
```

Comment:
`ORM relationship`, should mention that the object needs to relate back to a device.

Reply:
In thinking about it more, I think it is more confusing than it is worth. You can still accomplish the same thing, but put the onus on operator.

Something _like_ this fictitious `get_related_device` method.
`{{ interface["connected_circuit_termination"]["circuit"].get_related_device | get_secret("username") }}`

At the end of the day, however the user needs to get the device object returned, that is up to them. But honestly, it was confusing for me to understand, and it adds complexity I am not sure is worth it

Reply:
any object could be used with a custom `relationship`. For instance, a secret per `Circuit`.

Reply:
^^ That point is understood, but do you agree with my assessment?

Reply:
My understanding is that secrets can only apply to a Device. As an example, if you go to Device ObjectDetailView, you will see "Secret Group", but you will **not** see that in a circuit. There are many reasons to have Secrets associated to other "things" including other content/object types. That being said, there are other use cases we have not considered, but there are two concrete things we do know:

* You can request a secret, given a secret group Name/ID/Slug/etc 
* You can request a secret, given a Device, which will return the Secret Group

IMHO, the business logic one uses to come to the above two use cases, is beyond the scope of a plugin. The operator can build that logic theirselves in an endless amounts of interesting ways, but should we support this one use case? 

Also, I thinking about it there should be two filters, _roughly_ named:
* get_secret_by_device
* get_secret_by_secret_group

Reply:
I disagree with "secrets can only apply to a Device". In Nautobot Core, `Secrets` are only specifically used for `Device` model, but they can be used via `Relationships` in any other Nautobot model. And, I can see clear the use case for getting a specific secret key for a VPN tunnel, a Circuit, a BGP session, etc.
The current `get_secret` is the most generic one, that maybe we could expose in more "edible" ways, so following your suggestion, we could expose:
* `get_secret_by_secret_group`:  subset of the current `get_secret`
* `get_secret`: the current one, as the generic filter for secrets that could be linked to other object other than `Device`
* `get_secret_by_device`: assuming that the model is `Device`, valid for most of use cases